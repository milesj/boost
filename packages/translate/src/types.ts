import { i18n, InterpolationOptions } from 'i18next';

export type Direction = 'ltr' | 'rtl';

export type Format = 'js' | 'json' | 'yaml';

export type Locale = string;

export type InterpolationParams = Record<string, unknown>;

export interface MessageOptions {
	/** Default value to return if a translation was not found. */
	defaultValue?: string;
	/** Count used to determine plurals. */
	count?: number;
	/** Context used for special parsing (male, female, etc). */
	context?: string;
	/** Interpolation options to pass down. */
	interpolation?: InterpolationOptions;
	/** Force translation to this locale. */
	locale?: Locale;
	/** Post-processors to run on the translation. */
	postProcess?: string[] | string;
}

export interface Translator {
	direction: Direction;
	locale: Locale;
	changeLocale: (locale: Locale) => Promise<void>;
	(key: string[] | string, params?: InterpolationParams, options?: MessageOptions): string;
	// Testing only
	// eslint-disable-next-line @typescript-eslint/member-ordering
	i18n: i18n;
}
