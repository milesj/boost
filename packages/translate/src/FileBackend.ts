import fs from 'fs';
import path from 'path';
import i18next from 'i18next';
import { parseFile, Contract, Path, Predicates } from '@boost/common';
import { Locale, Format } from './types';

const EXTS: { [K in Format]: string[] } = {
  js: ['js'],
  json: ['json5', 'json'],
  yaml: ['yaml', 'yml'],
};

export interface FileBackendOptions {
  format?: Format;
  paths?: Path[];
}

export default class FileBackend extends Contract<FileBackendOptions>
  implements i18next.BackendModule {
  fileCache: { [path: string]: i18next.ResourceKey } = {};

  type: 'backend' = 'backend';

  init(services: unknown, options: Partial<FileBackendOptions>) {
    this.configure(options);

    // Validate resource paths are directories
    this.options.paths.forEach(resourcePath => {
      if (fs.existsSync(resourcePath) && !fs.statSync(resourcePath).isDirectory()) {
        throw new Error(`Resource path "${resourcePath}" must be a directory.`);
      }
    });
  }

  blueprint({ array, string }: Predicates) /* infer */ {
    return {
      format: string('yaml').oneOf(['js', 'json', 'yaml']),
      paths: array(string()),
    };
  }

  create() {
    // We don't need this but is required by the interface
  }

  read(
    locale: Locale,
    namespace: string,
    callback: (error: Error | null, resources: i18next.ResourceLanguage) => void,
  ): i18next.ResourceKey {
    const { format, paths } = this.options;
    const resources: i18next.ResourceKey = {};

    paths.forEach(resourcePath => {
      EXTS[format].some(ext => {
        const filePath = path.normalize(path.join(resourcePath, locale, `${namespace}.${ext}`));

        if (!this.fileCache[filePath] && !fs.existsSync(filePath)) {
          return false;
        }

        Object.assign(
          resources,
          this.fileCache[filePath] || (this.fileCache[filePath] = parseFile(filePath)),
        );

        return true;
      });
    });

    callback(null, resources);

    return resources;
  }
}
