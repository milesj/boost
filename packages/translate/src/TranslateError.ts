import { createScopedError } from '@boost/internal';

const errors = {
  LOCALE_REQUIRED: 'A locale must be defined if auto-detection is disabled.',
  NAMESPACE_REQUIRED: 'A namespace is required for translations.',
  RESOURCE_PATH_INVALID: 'Resource path "{0}" must be a directory.',
  RESOURCES_REQUIRED: 'At least 1 resource directory path is required.',
};

export type TranslateErrorCode = keyof typeof errors;

export default createScopedError<TranslateErrorCode>('TLT', 'TranslateError', errors);
