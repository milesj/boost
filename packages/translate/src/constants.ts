import factoryDebug from 'debug';

export const debug = factoryDebug('boost:translate');
