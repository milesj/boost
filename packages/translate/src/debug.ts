import { createInternalDebugger } from '@boost/internal';

export default createInternalDebugger('translate');
