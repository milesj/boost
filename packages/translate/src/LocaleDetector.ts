import type { LanguageDetectorModule } from 'i18next';
import { osLocaleSync } from 'os-locale';
import { debug } from './debug';
import type { Locale } from './types';

export class LocaleDetector implements LanguageDetectorModule {
	locale: Locale = 'en';

	type = 'languageDetector' as const;

	init() {
		// We don't need this but is required by the interface
	}

	cacheUserLanguage(locale: Locale) {
		this.locale = locale;
	}

	detect(): Locale {
		if (this.locale) {
			debug('Locale "%s" manually provided', this.locale);

			return this.locale;
		}

		return this.detectFromArgv() ?? this.detectFromOS();
	}

	detectFromArgv(): Locale | undefined {
		const args = process.argv;
		const index = args.indexOf('--locale');
		const nextIndex = index + 1;

		if (index >= 0 && args[nextIndex] && !args[nextIndex].startsWith('-')) {
			const locale = args[nextIndex];

			debug('Locale "%s" detected from --locale option', locale);

			return locale;
		}

		return undefined;
	}

	detectFromOS(): Locale {
		const locale = osLocaleSync().replace(/_/gu, '-');

		debug('Locale "%s" detected from operating system', locale);

		return locale;
	}
}
