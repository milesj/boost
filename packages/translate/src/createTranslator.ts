import i18next from 'i18next';
import { Path, toArray } from '@boost/common';
import LocaleDetector from './LocaleDetector';
import FileBackend from './FileBackend';
import { Locale, Translator, InterpolationParams } from './types';

// istanbul ignore next
function handleError(error: Error | null) {
  if (error) {
    throw error;
  }
}

export interface TranslatorOptions {
  debug?: boolean;
  fallbackLocale?: Locale;
  load?: i18next.InitOptions['load'];
  locale?: Locale;
}

export default function createTranslator(
  namespace: string | string[],
  resourcePath: Path | Path[],
  { debug, fallbackLocale, load, locale }: TranslatorOptions = {},
): Translator {
  const namespaces = toArray(namespace);
  const resourcePaths = toArray(resourcePath);

  if (namespaces.length === 0) {
    throw new Error('A namespace is required for translations.');
  } else if (resourcePaths.length === 0) {
    throw new Error('At least 1 resource file path is required.');
  }

  const translator = i18next
    .createInstance()
    .use(new LocaleDetector())
    .use(new FileBackend());

  translator.init(
    {
      backend: {
        resourcePaths,
      },
      debug,
      defaultNS: namespaces[0],
      fallbackLng: fallbackLocale || 'en',
      initImmediate: false,
      lng: locale,
      load,
      lowerCaseLng: true,
      ns: namespaces,
    },
    handleError,
  );

  function msg(key: string | string[], params?: InterpolationParams) {
    return translator.t(key, {
      interpolation: { escapeValue: false },
      replace: params,
    });
  }

  msg.dir = translator.dir();
  msg.locale = translator.language;

  msg.changeLocale = lang => {
    translator.changeLanguage(lang, error => {
      handleError(error);

      msg.dir = translator.dir();
      msg.locale = translator.language;
    });
  };

  return msg;
}
