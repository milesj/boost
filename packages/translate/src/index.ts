/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import createTranslator, { TranslatorOptions } from './createTranslator';
import TranslateError from './TranslateError';

export * from './types';

export { createTranslator, TranslatorOptions, TranslateError };
