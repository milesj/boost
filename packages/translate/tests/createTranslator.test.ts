import { getFixturePath } from '@boost/test-utils';
import createTranslator from '../src/createTranslator';
import { Translator } from '../src/types';

describe('createTranslator()', () => {
  let translator: Translator;

  beforeEach(() => {
    translator = createTranslator('common', getFixturePath('i18n-resources'));
  });

  it('errors if no namespace is provided', () => {
    expect(() => {
      createTranslator([], []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if no resource paths are provided', () => {
    expect(() => {
      createTranslator('common', []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if `autoDetect` and `locale` are empty', () => {
    expect(() => {
      createTranslator('common', getFixturePath('i18n-resources'), { autoDetect: false });
    }).toThrowErrorMatchingSnapshot();
  });

  it('returns a function to load translated messages', () => {
    expect(typeof translator).toBe('function');
    expect(translator('common:key')).toBe('value');
  });

  it('returns a message at the defined key', () => {
    expect(translator('common:key')).toBe('value');
  });

  it('sets primary namespace', () => {
    expect(translator('key')).toBe('value');
  });

  it('handles array of keys', () => {
    expect(translator(['unknown', 'key'])).toBe('value');
  });

  it('sets `dir` and `locale`', () => {
    expect(translator.direction).toBe('ltr');
    expect(translator.locale).toBe('en');
  });

  it('calls `t` on i18next instance', () => {
    const spy = jest.spyOn(translator.i18n, 't');

    expect(translator('missing', { foo: 'bar' }, { defaultValue: 'Hello' })).toBe('Hello');

    expect(spy).toHaveBeenCalledWith('missing', {
      defaultValue: 'Hello',
      interpolation: { escapeValue: false },
      lng: undefined,
      replace: { foo: 'bar' },
    });
  });

  describe('msg()', () => {
    beforeEach(() => {
      translator = createTranslator('common', [
        getFixturePath('i18n-resources'),
        getFixturePath('i18n-resources-more'),
      ]);
    });

    it('merges objects from multiple resource paths', () => {
      expect(translator('common:lang')).toBe('en');
      expect(translator('common:region')).toBe('region'); // MISSING
    });

    it('supports locale with region', async () => {
      await translator.changeLocale('en-US');

      expect(translator('common:lang')).toBe('en');
      expect(translator('common:region')).toBe('us');
    });
  });

  describe('changeLocale()', () => {
    it('calls `changeLanguage` on i18next', async () => {
      const spy = jest.spyOn(translator.i18n, 'changeLanguage');

      await translator.changeLocale('ja');

      expect(spy).toHaveBeenCalledWith('ja');
    });

    it('updates `dir` and `locale`', async () => {
      await translator.changeLocale('ja');

      expect(translator.direction).toBe('ltr');
      expect(translator.locale).toBe('ja');
    });
  });
});
