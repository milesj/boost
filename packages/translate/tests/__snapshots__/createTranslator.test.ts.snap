// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTranslator() errors if \`autoDetect\` and \`locale\` are empty 1`] = `"A locale must be defined if auto-detection is disabled. [TLT:LOCALE_REQUIRED]"`;

exports[`createTranslator() errors if no namespace is provided 1`] = `"A namespace is required for translations. [TLT:NAMESPACE_REQUIRED]"`;

exports[`createTranslator() errors if no resource paths are provided 1`] = `"At least 1 resource directory path is required. [TLT:RESOURCES_REQUIRED]"`;
